to-be-ccpp-ized/utilities/vertical_diffusion/vertical_diffusion_solver.F90

```diff
@@ -0,0 +1,136 @@
+module vertical_diffusion_solver
+
+    implicit none
+    private
+    save
+    
+    public :: fin_vol_solve
+    
+    contains
+    
+    ! Designed to solve the equation:
+    !
+    ! w * dq/dt = d/dp (D q' - v q) + c q
+    !
+    ! where q is a grid-cell average, and p is the vertical coordinate
+    ! (presumably pressure).
+    !
+    ! In this function, coef_q_weight == w, coef_q_diff == D,
+    ! coef_q_adv == v, and coef_q == c. All these are optional; omitting a
+    ! coefficient is equivalent to setting the entire array to 0.
+    !
+    ! coef_q_diff and coef_q_adv are defined at the level interfaces, while
+    ! coef_q and coef_q_weight are grid-cell averages.
+
+    function fin_vol_solve(dt, p, toSolve, ncols, pver, coef_q, coef_q_diff, coef_q_adv, &
+        coef_q_weight, upper_bndry, lower_bndry, l_cond, r_cond)  result(solution)
+    
+        use linear_1d_operators, only: &
+        zero_operator,               &
+        diagonal_operator,           &
+        diffusion_operator,          &
+        advection_operator,          &
+        BoundaryType,                &
+        TriDiagDecomp,               &
+        TriDiagOp,                   &
+        BoundaryCond,                &
+        operator(+)
+        use shr_kind_mod,        only: r8 => shr_kind_r8
+        use coords_1d,           only: Coords1D
+    
+        ! ---------------------- !
+        ! Input-Output Arguments !
+        ! ---------------------- !
+    
+        ! Time step.
+        real(r8), intent(in) :: dt
+        ! Grid spacings.
+        type(Coords1D), intent(in) :: p
+    
+        ! Matrix to decomp from.
+        ! real(r8), intent(in) :: u(ncols,pver)
+        integer,  intent(in)    :: ncols
+        integer,  intent(in)    :: pver
+        real(r8), intent(in)    :: toSolve(ncols,pver)
+    
+        ! Coefficients for diffusion and advection.
+        !
+        ! The sizes must be consistent among all the coefficients that are
+        ! actually present, i.e. coef_q_diff and coef_q_adv should be one level
+        ! bigger than coef_q and coef_q_weight, and have the same column number.
+        real(r8), contiguous, intent(in), optional :: coef_q(:,:), &
```

Comment:
Here `contiguous` is used directly but the CESM-specific CPP macro `USE_CONTIGUOUS` is used in `coords_1d.F90` and `linear_1d_operators.F90`. Maybe the same macro could be used here?

Reply:
I think the old `USE_CONTIGUOUS` macro is not needed and should be discontinued. I think this harks back to a time when not all Fortran compilers supported the `contiguous` attribute.

Reply:
Since the solve module is a new wrapper around the solve step and the other 2 files are helper files being copied directly over from CAM, is it reasonable to leave `linear_1d_operators` and `coords_1d` as is and prefer `contiguous` in new/refactored code moving forward?

Reply:
I would personally vote for now to do whatever is easiest.  I can say that fully CCPP-ized physics routines will have to use the `contiguous` atribute, as `USE_CONTIGUOUS` is CESM-specific and thus not currently portable.

Reply:
Perfect, I'm happy to leave it as is then.

Reply:
Thanks @gold2718, @nusbaume, @mwaxmonsky, marking as resolved.